server/handlers/clients.js

```diff
@@ -0,0 +1,256 @@
+    logger.debug('body:', body);
+
+    // proxy the token
+    token = client.token;
+    if ('token' in payload) {
+      client.token = payload.token;
+      logger.debug('Using custom client token');
+    }
+    const response = await client.request(method, path, payload);
+    logger.debug('server response:', response ? response : 'null');
+    if (response) return res.status(200).json(response);
+    // return 404 when response is null due to
+    // resource not being found on server
+    return res.status(404).json({ error: { message: 'resource not found' } });
+  } catch (e) {
+    logger.error(`Error querying ${client.constructor.name}:`, e);
+    return res
+      .status(502)
+      .send({ error: { message: e.message, code: e.code, type: e.type } });
+  } finally {
+    // always reassign the original token
+    client.token = token;
+  }
+}
+
+async function getConfigHandler(req, res) {
+  const { logger } = req;
+  let configurations;
+  try {
+    configurations = await getConfig(req.params.id);
+  } catch (e) {
+    logger.error(e);
+    if (e.code === 'ENOENT')
+      return res.status(404).json({
+        error: {
+          message: `Config for "${req.params.id}" not found`,
+          code: 404
+        }
+      });
+    else
+      return res.status(500).json({
+        error: {
+          message: `There was a problem with your request.`,
+          code: 500
+        }
+      });
+  }
+
+  const info = {
+    configs: configurations.data
+  };
+
+  if (req.query.health) {
+    try {
+      logger.info(`Checking status of client "${req.params.id}"...`);
+      const [err, clientErrors] = await testConfigOptions(configurations);
+      if (!err) info.healthy = true;
+      else if (clientErrors.failed.length) {
+        info.failed = clientErrors.failed;
+        info.healthy = false;
+      }
+    } catch (e) {
+      return res.status(500).send(e);
+    }
+  }
+
+  // scrub apiKeys and tokens
+  for (let key in configurations.data) {
+    if (key.includes('api') || key.includes('token'))
+      configurations.data[key] = undefined;
+  }
+
+  res.status(200).json(info);
+}
+
+function addConfigHandler(req, res) {
+  const { config } = req;
+  const id = req.params.id;
+
+  if (!configsMap) {
+    const clientConfigs = loadClientConfigs(config);
+    configsMap = createConfigsMap(clientConfigs);
+  }
+  if (configsMap.get(id))
+    return res
+      .status(409)
+      .send({ message: `A client with the id '${id}' already exists` });
+
+  return updateOrAdd(req, res);
+}
+
+function updateConfigHandler(req, res) {
+  return updateOrAdd(req, res);
+}
+
+function deleteConfigHandler(req, res) {
+  const success = deleteConfig(req.params.id);
+  return res.status(200).json({ success });
+}
+
+async function updateOrAdd(req, res) {
+  const { logger } = req;
+  const id = req.params.id;
+  try {
+    const { options, force = false } = req.body;
+    const config = await createClientConfig(id, options, JSON.parse(force));
```

Comment:
Using `JSON.parse` could potentially by a DOS vector, imagine if `force` is a hugely nested object. Maybe we want a middleware that rejects requests that are too large

Reply:
you can type cast explicitly using `Boolean(force)` or js shorthand `!!force`
This will catch other values though so maybe just checking for the strings is a better idea

Reply:
added some explicit checks for this now warns if it's not `true` or `false` (bool or string)